pkg/container/versioned/value.go

```diff
@@ -0,0 +1,510 @@
+
+func newVersionHandle(version version, coordinator *Coordinator) *VersionHandle {
+	// handle on maxVersion never expires
+	if version == maxVersion {
+		coordinator = nil
+	}
+	h := &VersionHandle{version: version}
+	h.coordinator.Store(coordinator)
+	if coordinator != nil {
+		// Set a finalizer to catch unclosed handles. The finalizer function complains
+		// loudly, so that we do not rely the finalizer for closing.
+		runtime.SetFinalizer(h, versionHandleFinalizer)
+	}
+	return h
+}
+
+// Latest returns a VersionHandle for the latest version of current/non-removed values
+// Only to be used in cases where the latest values are needed and no transactionality is required.
+func Latest() *VersionHandle {
+	return newVersionHandle(maxVersion, nil)
+}
+
+type atomicVersion struct {
+	version atomic.Uint64
+}
+
+func (a *atomicVersion) load() version {
+	return version(a.version.Load())
+}
+
+func (a *atomicVersion) store(version version) {
+	a.version.Store(uint64(version))
+}
+
+type versionCount struct {
+	version version
+	count   int
+}
+
+// Coordinator defines a common version number space for multiple versioned.Values,
+// and provides facilities for cleaning out stale versions.
+// The Value's are not directly managed by the Coordinator, but all the values under coordination
+// should be cleaned by the 'cleaner' function given to the Coordinator.
+type Coordinator struct {
+	// Logger supplied to NewCoordinator. Should be set if logging is desired.
+	Logger *logrus.Entry
+
+	// Cleaner is called with the earliest version that must be kept.
+	// Must be set to clean up resources held for old versions.
+	// Cleaner function may be called concurrently, the function must synchronize
+	// use of any shared resources.
+	Cleaner func(KeepVersion)
+
+	// mutex protects the rest of the fields
+	mutex lock.RWMutex
+
+	// version is the version number of the last applied change
+	version version
+
+	// oldest version not cleaned off
+	oldestVersion version
+
+	// versions is an ordered list of outstanding VersionHandles with a reference count.
+	// Outdated values can be cleaned when there are no outstanding VersionHandles for them.
+	versions []versionCount
+}
+
+// PrepareNextVersion returns a transaction to be used when adding or removing values.
+//
+// Callers need to coordinate so that a single goroutine is performing modifications at any one
+// time, consisting of the following operations:
+//
+// - tx := coordinator.PrepareNextVersion()
+//   - value.SetAt(... , tx)
+//   - value.RemoveAt(..., tx)
+//   - ...
+//
+// - tx.Commit()
+func (v *Coordinator) PrepareNextVersion() *Tx {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	return &Tx{
+		nextVersion: v.version + 1,
+		coordinator: v,
+	}
+}
+
+// All values in Tx are constants
+type Tx struct {
+	nextVersion version
+	coordinator *Coordinator
+}
+
+// LatestTx refers to maxVersion without having a coordinator, should only be used for testing.
+var LatestTx = &Tx{nextVersion: maxVersion}
```

Comment:
nit: if you move this to the _test.go file, it will only be possible to use it in tests.

Reply:
But only from the `versioned` package itself. This is used also from `pkg/policy`.